### PR TITLE
[Feature] Implementation of ldexp

### PR DESCRIPTION
**Pull request by GDBobby**
Return of ldexp

### CODE REVIEW

**Comment by GDBobby** [comment]
bits.hpp line 81
const auto bits = bit_cast<ccm::helpers::float_bits_t<T>>(x);
bit_cast is returning an incorrect value when the type is double and equal to negative zero
the correct value should be 0x8000000000000000
the value returned is 0xCCCCCCCC00000001
When the result of ldexp is casted to a float it's still correct, but I'm assuming that's unacceptable

File: include/ccmath/math/fmanip/ldexp.hpp [hunk 1]
```diff
@@ -26,8 +26,9 @@ namespace ccm
 		int32_t oldexp = ccm::support::get_exponent_of_floating_point<Floating_Point>(num);
```
Comment (Rinzii) [inline include/ccmath/math/fmanip/ldexp.hpp:26]: Use std::int32_t instead of int32_t without the std namespace.

File: include/ccmath/math/exponential/impl/exp2_data.hpp [hunk 2]
```diff
@@ -52,7 +52,7 @@ namespace ccm::internal
 		static constexpr std::size_t k_exp2_poly_order_dbl = 5;
 
 	private:
-		static constexpr std::size_t internal_table_size = static_cast<const std::size_t>(2 * (1 << k_exp2_table_bits_dbl));
+		static constexpr std::size_t internal_table_size = (std::size_t)(2 * (1 << k_exp2_table_bits_dbl));
```
Comment (Rinzii) [inline include/ccmath/math/exponential/impl/exp2_data.hpp:55]: Ah this is a bug. Change this instead to `static_cast<std::size_t>` and avoid using C style casts.

File: include/ccmath/math/fmanip/ldexp.hpp [hunk 3]
```diff
@@ -8,7 +8,108 @@
...
+	 * @brief Returns the floating-point remainder of the division operation x/y.
+	 * @note
+	 * @tparam num A floating-point type.
+	 * @param exp An integer value.
+	 * @return The result of parameter num multiplied by 2 to the power of parameter exp.
+	 */
+	inline constexpr long double ldexpl(long double num, int exp) noexcept
+	{
+		//long double isn't supported yet
+		return std::numeric_limits<long double>::signaling_NaN(); // ldexp<long double>(num, exp);
```
Comment (Rinzii) [inline include/ccmath/math/fmanip/ldexp.hpp:112]: We should be instead casting long double to double and passing it into our function instead of returning a signaling NaN. This is a more desirable outcome than the latter.

File: test/fmanip/ldexp_test.cpp [hunk 4]
```diff
@@ -14,6 +14,35 @@
...
+	EXPECT_DOUBLE_EQ(std::ldexp(5.0, 4), ccm::ldexp(5.0, 4));
+	EXPECT_DOUBLE_EQ(std::ldexp(-0.0, 1), ccm::ldexp(-0.0, 1));
+	EXPECT_DOUBLE_EQ(std::ldexp(0.0, 1), ccm::ldexp(0.0, 1));
+	EXPECT_FLOAT_EQ(std::ldexp(5.f, -4), ccm::ldexp(5.f, -4));
+	EXPECT_DOUBLE_EQ(std::ldexp(-5.0, -4), ccm::ldexp(-5.0, -4));
+	EXPECT_FLOAT_EQ(std::ldexp(720.32f, 22), ccm::ldexp(720.32f, 22));
+	EXPECT_DOUBLE_EQ(std::ldexp(1, -1074), ccm::ldexp(1, -1074));
+	EXPECT_DOUBLE_EQ(std::ldexp(1, 1024), ccm::ldexp(1, 1024));
+
+	EXPECT_FLOAT_EQ(std::ldexp(std::numeric_limits<float>::max(), std::numeric_limits<int>::max()),
```
Comment (Rinzii) [inline test/fmanip/ldexp_test.cpp:28]: We should also be testing for edge cases with double not just float.

**Review by Rinzii — CHANGES_REQUESTED** [review]
First I want to thank you again for applying the requested changes! So far everything looks pretty good and I only have some minor requests.

File: include/ccmath/math/fmanip/ldexp.hpp [hunk 5]
```diff
@@ -8,7 +8,108 @@
...
 {
+	/**
+	 * @brief Multiplies a floating point value num by the number 2 raised to the exp power.
+	 * @note On many implementations, std::ldexp is less efficient than multiplication or division by a power of two using arithmetic operators.
+	 * @tparam num A floating-point type.
+	 * @param exp An integer value.
+	 * @return If no errors occur, num multiplied by 2 to the power of exp (num×2exp) is returned.\nIf a range error due to overflow occurs, ±HUGE_VAL, ±HUGE_VALF, or ±HUGE_VALL is returned.\nIf a range error due to underflow occurs, the correct result (after rounding) is returned.
+	 */
+	template <typename Floating_Point, std::enable_if_t<!std::is_integral_v<Floating_Point>, bool> = true>
+	inline constexpr Floating_Point ldexp(Floating_Point num, std::int32_t exp) noexcept
```
Comment (Rinzii) [inline include/ccmath/math/fmanip/ldexp.hpp:24]: exp should be int here to properly mimic the api of the standard.

File: include/ccmath/math/fmanip/ldexp.hpp [hunk 6]
```diff
@@ -8,7 +8,108 @@
...
 namespace ccm
 {
+	/**
+	 * @brief Multiplies a floating point value num by the number 2 raised to the exp power.
+	 * @note On many implementations, std::ldexp is less efficient than multiplication or division by a power of two using arithmetic operators.
+	 * @tparam num A floating-point type.
+	 * @param exp An integer value.
+	 * @return If no errors occur, num multiplied by 2 to the power of exp (num×2exp) is returned.\nIf a range error due to overflow occurs, ±HUGE_VAL, ±HUGE_VALF, or ±HUGE_VALL is returned.\nIf a range error due to underflow occurs, the correct result (after rounding) is returned.
+	 */
+	template <typename Floating_Point, std::enable_if_t<!std::is_integral_v<Floating_Point>, bool> = true>
```
Comment (Rinzii) [inline include/ccmath/math/fmanip/ldexp.hpp:23]: Change the typename here to just "T" to mimic the naming conventions of ccmath.

File: test/fmanip/ldexp_test.cpp [hunk 7]
```diff
@@ -14,6 +14,60 @@
...
+	EXPECT_DOUBLE_EQ(std::ldexp(std::numeric_limits<double>::max(), std::numeric_limits<int>::min()),
+					ccm::ldexp(std::numeric_limits<double>::max(), std::numeric_limits<int>::min()));
+	EXPECT_FLOAT_EQ(std::ldexp(std::numeric_limits<float>::max(), std::numeric_limits<int>::min()),
+					ccm::ldexp(std::numeric_limits<float>::max(), std::numeric_limits<int>::min()));
+
+	EXPECT_DOUBLE_EQ(ccm::ldexp(std::numeric_limits<double>::infinity(), 10), std::ldexp(std::numeric_limits<double>::infinity(), 10));
+	EXPECT_FLOAT_EQ(ccm::ldexp(std::numeric_limits<float>::infinity(), 10), std::ldexp(std::numeric_limits<float>::infinity(), 10));
+
+	// Test for edge case where if either argument is NaN, NaN is returned.
+	auto testForNanCcmIfEitherArgumentIsNand	   = std::isnan(ccm::ldexp(std::numeric_limits<double>::quiet_NaN(), 10));
```
Comment (Rinzii) [inline test/fmanip/ldexp_test.cpp:64]: We should also be testing for if exp is passed qNaN and num is an expected value.

File: test/fmanip/ldexp_test.cpp [hunk 8]
```diff
@@ -14,6 +14,60 @@
...
+					ccm::ldexp(std::numeric_limits<double>::min(), std::numeric_limits<int>::max()));
+	EXPECT_FLOAT_EQ(std::ldexp(std::numeric_limits<float>::min(), std::numeric_limits<int>::max()),
+					ccm::ldexp(std::numeric_limits<float>::min(), std::numeric_limits<int>::max()));
+
+	EXPECT_DOUBLE_EQ(std::ldexp(std::numeric_limits<double>::max(), std::numeric_limits<int>::min()),
+					ccm::ldexp(std::numeric_limits<double>::max(), std::numeric_limits<int>::min()));
+	EXPECT_FLOAT_EQ(std::ldexp(std::numeric_limits<float>::max(), std::numeric_limits<int>::min()),
+					ccm::ldexp(std::numeric_limits<float>::max(), std::numeric_limits<int>::min()));
+
+	EXPECT_DOUBLE_EQ(ccm::ldexp(std::numeric_limits<double>::infinity(), 10), std::ldexp(std::numeric_limits<double>::infinity(), 10));
```
Comment (Rinzii) [inline test/fmanip/ldexp_test.cpp:60]: We should also be testing for if exp is passed Inf and num is an expected value.

File: include/ccmath/math/fmanip/ldexp.hpp [hunk 9]
```diff
@@ -8,7 +8,108 @@
...
+	 * @note On many implementations, std::ldexp is less efficient than multiplication or division by a power of two using arithmetic operators.
+	 * @tparam num An integer type.
+	 * @param exp An integer value.
+	 * @return If no errors occur, num multiplied by 2 to the power of exp (num×2exp) is returned.\nIf a range error due to overflow occurs, ±HUGE_VAL,
+	 * ±HUGE_VALF, or ±HUGE_VALL is returned.\nIf a range error due to underflow occurs, the correct result (after rounding) is returned.
+	 */
+	template <typename Integer, std::enable_if_t<std::is_integral_v<Integer>, bool> = true>
+	inline constexpr double ldexp(Integer num, std::int32_t exp) noexcept
+	{
+		return ldexp<double>(static_cast<double>(num), exp);
```
Comment (Rinzii) [inline include/ccmath/math/fmanip/ldexp.hpp:89]: Apply the ccm namespace to all functions that are calling into the top function.

**Review by Rinzii — CHANGES_REQUESTED** [review]
Once again thank you for your changes.

These should be my last requests and there all pretty minor!

**Comment by GDBobby** [comment]
exp is an int

**Comment by Rinzii** [comment]
> exp is an int

My bad must of been shown an older bit of code when I was doing my review.

**Comment by Rinzii** [comment]
Also remember to press the "Resolve conversation" button once a change request has been done! This helps me know if everything has been resolved. ^^

**Comment by GDBobby** [comment]
My bad, didn't see the merge conflicts. Having a multitude of unrelated issues to that now lol

**Review by Rinzii — APPROVED** [review]
Everything looks good to me! :tada:
